### PR TITLE
float16 training for resnet

### PR DESCRIPTION
**Pull request by ap-hynninen**
Adds float16 training for cudnn batch_norm. Introduces example resnet_fp16.py script for float16 training of the resnet model. Tested for convergence on resnet-50 by comparing results with float32 training.

### CODE REVIEW

**Comment by piiswrong** [comment]
@winstywang

**Comment by piiswrong** [comment]
not so sure about the different type between param and aux. Seems confusing...

**Comment by ap-hynninen** [comment]
These are dictated by cudnn

**Comment by ap-hynninen** [comment]
x and y are in fp16, rest are in fp32. This is a little hard to gather from cudnn documentation. It is mentioned in the chapter about "cudnnDeriveBNTensorDescriptor"

**Comment by piiswrong** [comment]
please fix compile error

File: src/operator/cudnn_batch_norm-inl.h [hunk 1]
```diff
@@ -158,71 +168,79 @@ class CuDNNBatchNormOp : public Operator {
...
-
-    if (param_.fix_gamma) gamma = 1.f;
+    Tensor<gpu, 4, DType> x =
+      in_data[cudnnbatchnorm::kData].get_with_shape<gpu, 4, DType>(shape_, s);
+    Tensor<gpu, 4, DType> dx =
+      in_grad[cudnnbatchnorm::kData].get_with_shape<gpu, 4, DType>(shape_, s);
+    Tensor<gpu, 4, DType> dy =
+      out_grad[cudnnbatchnorm::kOut].get_with_shape<gpu, 4, DType>(shape_, s);
+
+    MSHADOW_REAL_TYPE_SWITCH(dtype_param_, DTypeParam, {
```
Comment (sxjscience) [inline src/operator/cudnn_batch_norm-inl.h]: I find that the windows build fails due to the nested macro here.

Comment (ap-hynninen) [inline src/operator/cudnn_batch_norm-inl.h]: Argh... I'll try this tomorrow on Windows.

Comment (sxjscience) [inline src/operator/cudnn_batch_norm-inl.h]: I've fixed here https://github.com/dmlc/mxnet/pull/5352, should be ok now.

**Pull request merged**